### PR TITLE
Proposal: Remove side-effects from hook package

### PR DESCRIPTION
**Pull request by moorereason**
This commit removes logging from the hook package and relies on returning errors
to convey...errors.  Encountered errors are returned immediately.  This commit
may alter the behavior of hook.  If errors were logged in the past but the given
function did not return immediately, this commit would change that behavior.

Tests have been updated to validate error return values (ie. err == nil) and
provide test coverage for a few new lines of code introduced by this commit.

Most of the testing was done via unit testing in the hook package.  Changes to
webhook.go were not fully tested.  Beware.

### CODE REVIEW

**Comment by moorereason** [comment]
Curious about your thoughts.

This PR would resolve #26.

**Comment by adnanh** [comment]
I like the idea and I'll review this as soon as I get some time. I have a lot of going on at the moment :-)

**Comment by sebest** [comment]
Hello @moorereason ,

I think it is a great idea to return errors, but to make them really useful (more than just testing != nil) you should either define all of them , so we can compare them.
see listing 1.10 here: http://www.goinggo.net/2014/10/error-handling-in-go-part-i.html

Or use/implement the error interface if you want variable error message
see http://www.goinggo.net/2014/11/error-handling-in-go-part-ii.html

**Comment by moorereason** [comment]
Good point, @sebest.  I'll take a look at doing that when I get some free time.  Thanks

**Comment by moorereason** [comment]
Rebased my original commit and added a new commit that uses custom error types.  Let me know what you think.

File: webhook.go [hunk 1]
```diff
@@ -197,13 +197,24 @@ func hookHandler(w http.ResponseWriter, r *http.Request) {
 }
 
 func handleHook(hook *hook.Hook, headers, query, payload *map[string]interface{}, body *[]byte) {
-	if hook.TriggerRule == nil || hook.TriggerRule != nil && hook.TriggerRule.Evaluate(headers, query, payload, body) {
+	if hook.TriggerRule == nil {
+		log.Printf("%q hook did not get triggered: no defined trigger rule\n", hook.ID)
+		return
```
Comment (adnanh) [inline webhook.go]: Hook should always trigger if the rule isn't specified

Comment (moorereason) [inline webhook.go]: Fixed

**Comment by moorereason** [comment]
Adnan,
Care to comment on why you closed this PR?  Did you find any issues with it?

I rebased this PR against the latest *master* branch on Friday, but I still need to do some more testing.  I pushed up my changes to my *sidefx* branch so that I could work on it from home over the weekend (but I never got around to it).

If you would prefer, I can delete this PR and submit a new one once I've finished testing against the latest *master* branch.  Or would you prefer that I submit PRs to the *development* branch?

**Comment by adnanh** [comment]
Yes,  I wanted to ask you to submit a new PR to development branch :)
